sstables/sstables.cc

```diff
@@ -3365,6 +3365,30 @@ future<> remove_table_directory_if_has_no_snapshots(fs::path table_dir) {
     }
 }
 
+bool sstable::needs_cleanup(const dht::token_range_vector& sorted_owned_ranges, schema_ptr s) const {
+    auto first_token = get_first_decorated_key().token();
+    auto last_token = get_last_decorated_key().token();
+    dht::token_range sst_token_range = dht::token_range::make(first_token, last_token);
+
+    auto r = std::lower_bound(sorted_owned_ranges.begin(), sorted_owned_ranges.end(), first_token,
+            [] (const range<dht::token>& a, const dht::token& b) {
+        // check that range a is before token b.
+        return a.after(b, dht::token_comparator());
+    });
+
+    // return true iff sst partition range isn't fully contained in any of the owned ranges.
+    if (r != sorted_owned_ranges.end()) {
+        if (r->contains(sst_token_range, dht::token_comparator())) {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool sstable::mark_for_cleanup(const dht::token_range_vector& sorted_owned_ranges, schema_ptr s) {
+    return _requires_cleanup = needs_cleanup(sorted_owned_ranges, std::move(s));
+}
+
```

Comment:
Ideally, we don't want to embed distributed knowledge into storage layer, as it's a layer violation. If compaction manager knows which sstables require cleanup, perhaps we don't need sstable itself knowing it needs cleanup, but rather push that decision to compaction manager, which is cleanup aware.

Reply:
we call it from reshard as well, so it's not only compaction_manager but we can expose compaction_manager::needs_cleanup() so it can be used elsewhere.

Reply:
Makes sense.

Reply:
I see @raphaelsc had the same idea about mark_for_cleanup() as I did.